### ISSUE TITLE
Error when URL parameter is of type array

### ISSUE THREAD

**Issue by stimpy77** [rbeauchamp/Swashbuckle.OData#23]
I'll try to follow up with a dummy sample gist but I'm getting this error while the Swashbuckle.OData provider is enabled.

	<Error><Message>An error has occurred.</Message><ExceptionMessage>Could not generate sample value for query parameter type array and format null</ExceptionMessage><ExceptionType>System.Exception</ExceptionType><StackTrace>   at Swashbuckle.OData.ODataApiExplorer.GenerateSampleQueryParameterValue(Parameter queryParameter)
	   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
	   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector)
	   at Swashbuckle.OData.ODataApiExplorer.GenerateSampleQueryParameterValues(Operation operation)
	   at Swashbuckle.OData.ODataApiExplorer.GenerateSampleODataAbsoluteUri(String pathTemplate, Operation operation)
	   at Swashbuckle.OData.ODataApiExplorer.GetApiDescription(HttpMethod httpMethod, ODataRoute oDataRoute, String potentialPathTemplate, Operation potentialOperation)
	   at Swashbuckle.OData.ODataApiExplorer.GetApiDescriptions(ODataRoute oDataRoute, String potentialPathTemplate, PathItem potentialOperations)
	   at Swashbuckle.OData.ODataApiExplorer.GetApiDescriptions(ODataRoute oDataRoute)
	   at Swashbuckle.OData.ODataApiExplorer.GetApiDescriptions()
	   at System.Lazy`1.CreateValue()
	   at System.Lazy`1.LazyInitValue()
	   at System.Lazy`1.get_Value()
	   at Swashbuckle.OData.ODataApiExplorer.get_ApiDescriptions()
	   at Swashbuckle.OData.ODataSwaggerProvider.GetApiDescriptionsFor(String apiVersion)
	   at Swashbuckle.OData.ODataSwaggerProvider.GetSwagger(String rootUrl, String apiVersion)
	   at Swashbuckle.Application.SwaggerDocsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	   at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	   at System.Web.Http.Dispatcher.HttpRoutingDispatcher.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	   at System.Net.Http.DelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	   at System.Web.Http.Cors.CorsMessageHandler.&lt;&gt;n__FabricatedMethod3(HttpRequestMessage , CancellationToken )
	   at System.Web.Http.Cors.CorsMessageHandler.&lt;SendAsync&gt;d__0.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
	   at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
	   at System.Web.Http.Owin.PassiveAuthenticationMessageHandler.&lt;SendAsync&gt;d__0.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
	   at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
	   at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
	   at System.Web.Http.HttpServer.&lt;SendAsync&gt;d__0.MoveNext()</StackTrace></Error>

**Comment by rbeauchamp** [comment]
Hey @stimpy77, I'll re-open after you provide more detail. Thanks!

**Comment by stimpy77** [comment]
It's caused by our having a custom route. Still want to provide a gist but perhaps it's easily repro'd by just declaring a custom route.

**Comment by rbeauchamp-lendingtree** [comment]
Looks like it has something to do with a route that takes an array of URL parameters. Do you see such a route?

**Comment by stimpy77** [comment]
Yeah I found out from my lead that he found that array types aren't even handled, it was a one-liner in a file I don't know what he was fussing with but it was a list of type handlers in either one of the libs (swashbuckle or *.odata), don't know which. Sorry, I still don't have gist for you, I've just been busy.

**Comment by stimpy77** [comment]
It was in ODataApiExplorer.cs, under case null:, switch (type), add case "array": return "[]";

**Comment by stimpy77** [comment]
Sorry, the context of that last comment is in Swashbuckle.OData.Descriptions.ParameterExtensions in GenerateSamplePathParameterValue(), not ODataApiExplorer

**Comment by rbeauchamp-lendingtree** [comment]
Thanks @stimpy77 !

**Comment by rbeauchamp** [comment]
@stimpy77 Fix is available in [v2.7.2](https://www.nuget.org/packages/Swashbuckle.OData/2.7.2). Please try it out and let me know if you find any issues. Thanks!

**Comment by stimpy77** [comment]
Thanks. I dunno. *NullReferenceException*

       at Swashbuckle.OData.Descriptions.ParameterExtensions.GetEntityTypeForBodyParameter(Parameter parameter)
       at Swashbuckle.OData.Descriptions.ParameterExtensions.GetClrType(Parameter parameter)
       at Swashbuckle.OData.Descriptions.MapToDefault.Map(Parameter swaggerParameter, Int32 parameterIndex, HttpActionDescriptor actionDescriptor)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.&lt;&gt;c__DisplayClass6_0.&lt;GetHttpParameterDescriptor&gt;b__0(IParameterMapper mapper)
       at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
       at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.GetHttpParameterDescriptor(Parameter parameter, Int32 index, HttpActionDescriptor actionDescriptor)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.GetParameterDescription(Parameter parameter, Int32 index, HttpActionDescriptor actionDescriptor)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.&lt;&gt;c__DisplayClass4_0.&lt;CreateParameterDescriptions&gt;b__0(Parameter parameter, Int32 index)
       at System.Linq.Enumerable.&lt;SelectIterator&gt;d__5`2.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.CreateParameterDescriptions(Operation operation, HttpActionDescriptor actionDescriptor)
       at Swashbuckle.OData.Descriptions.SwaggerOperationMapper.Map(ODataActionDescriptor oDataActionDescriptor)
       at Swashbuckle.OData.Descriptions.ODataApiExplorer.&lt;&gt;c__DisplayClass7_0.&lt;GetApiDescriptions&gt;b__2(IODataActionDescriptorMapper mapper)
       at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
       at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
       at Swashbuckle.OData.Descriptions.ODataApiExplorer.&lt;GetApiDescriptions&gt;b__7_1(ODataActionDescriptor oDataActionDescriptor)
       at System.Linq.Enumerable.&lt;SelectManyIterator&gt;d__16`2.MoveNext()
       at System.Linq.Enumerable.&lt;DistinctIterator&gt;d__63`1.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at Swashbuckle.OData.CollectionExtentions.ToCollection[T](IEnumerable`1 source)
       at Swashbuckle.OData.Descriptions.ODataApiExplorer.GetApiDescriptions()
       at System.Lazy`1.CreateValue()
       at System.Lazy`1.LazyInitValue()
       at System.Lazy`1.get_Value()
       at Swashbuckle.OData.Descriptions.ODataApiExplorer.get_ApiDescriptions()
       at Swashbuckle.OData.ODataSwaggerProvider.GetApiDescriptionsFor(String apiVersion)
       at Swashbuckle.OData.ODataSwaggerProvider.GetSwagger(String rootUrl, String apiVersion)
       at Swashbuckle.Application.SwaggerDocsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
       at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
       at System.Web.Http.Dispatcher.HttpRoutingDispatcher.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
       at System.Net.Http.DelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
       at System.Web.Http.Cors.CorsMessageHandler.&lt;&gt;n__FabricatedMethod3(HttpRequestMessage , CancellationToken )
       at System.Web.Http.Cors.CorsMessageHandler.&lt;SendAsync&gt;d__0.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at System.Web.Http.Owin.PassiveAuthenticationMessageHandler.&lt;SendAsync&gt;d__0.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at System.Web.Http.Tracing.ITraceWriterExtensions.&lt;TraceBeginEndAsyncCore&gt;d__18`1.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at System.Web.Http.HttpServer.&lt;SendAsync&gt;d__0.MoveNext()

**Comment by rbeauchamp** [comment]
Hey Jon, So I understand: you added null propagation to Swashbuckle.OData to fix this issue?

**Comment by stimpy77** [comment]
The particular method where the exception is occurring has a single line of code that is the culprit, and one of the properties it is querying a property of is null. I added null propagation, which moved the exception downstream. I followed this downstream further at least a couple times until the exceptions went away. But I still don't know what this particular function was reading.

**Comment by rbeauchamp** [comment]
any chance can you forward the changes you made? :) Could help even if we don't know the particular cause. Thanks.

**Comment by stimpy77** [comment]
sorry, it wasn't a fork, I just copied all the .cs files from the swashbuckle.odata and swashbuckle projects directly into our project (rather than pull the project in) so that we didn't have nuget dependencies, so I can't diff at this point, and even if I could I had to strip out all the C#6 syntax as we're using a C#5 compiler in TFS.

I'm sorry I'm not as immediately helpful as I should be. I might follow up though.

**Comment by stimpy77** [comment]
sorry, it wasn't a fork, I just copied all the .cs files from the swashbuckle.odata and swashbuckle projects directly into our project (rather than pull the project in) so that we didn't have nuget dependencies, so I can't diff at this point, and even if I could I had to strip out all the C#6 syntax as we're using a C#5 compiler in TFS. Since I put everything in the same project I quickly lost track of my changes, I just wanted to get a complete and valid .json emitted as quickly as possible.

I'm sorry I'm not as immediately helpful as I should be. I might follow up though. FWIW, here's the Swagger and Swagger.OData combined I have (attached).
[Swashbuckle.zip](https://github.com/rbeauchamp/Swashbuckle.OData/files/72975/Swashbuckle.zip)
